src/ump/api/routes/ensembles.py

```diff
@@ -0,0 +1,128 @@
+    if auth is None:
+        return Response('[]', mimetype = 'application/json')
+    with Session(engine) as session:
+        stmt = select(Ensemble).where(Ensemble.user_id == auth['sub'])
+        return jsonify(session.scalars(stmt).fetchall())
+
+@ensembles.route("/", methods=["POST"])
+def create():
+    auth = g.get('auth_token')
+    if auth is None:
+        logging.error('Not creating ensemble, no authentication found.')
+        return Response('', mimetype = 'application/json')
+    data = request.get_json()
+    ensemble = Ensemble(
+        name = data['name'],
+        description = data['description'],
+        user_id = auth['sub'],
+        scenario_configs = json.dumps(data['scenario_configs'])
+    )
+    with Session(engine) as session:
+        session.add(ensemble)
+        session.commit()
+        return Response(mimetype="application/json", status = 201)
+
+@ensembles.route('/<path:ensemble_id>/comments', methods = ['GET'])
+def get_comments(ensemble_id):
+    auth = g.get('auth_token')
+    if auth is None:
+        return Response('[]', mimetype = 'application/json')
+    with Session(engine) as session:
+        stmt = select(Comment).where(Comment.user_id == auth['sub']).where(Comment.ensemble_id == ensemble_id)
+        return jsonify(session.scalars(stmt).fetchall())
+
+@ensembles.route("/<path:ensemble_id>/comments", methods=["POST"])
+def create_comment(ensemble_id):
+    auth = g.get('auth_token')
+    if auth is None:
+        logging.error('Not creating comment, no authentication found.')
+        return Response('', mimetype = 'application/json')
+    comment = Comment(
+        user_id = auth['sub'],
+        ensemble_id = ensemble_id,
+        comment = request.get_json()['comment']
+    )
+    with Session(engine) as session:
+        session.add(comment)
+        session.commit()
+        return Response(mimetype="application/json", status = 201)
+
+@ensembles.route("/<path:ensemble_id>", methods=["GET"])
+def get(ensemble_id):
+    auth = g.get('auth_token')
+    if auth is None:
+        return Response('[]', mimetype = 'application/json')
+    with Session(engine) as session:
+        stmt = select(Ensemble).where(Ensemble.user_id == auth['sub']).where(Ensemble.id == ensemble_id)
+        return jsonify(session.scalars(stmt).fetchall())
+    return Response(json.dumps(''), mimetype="application/json")
+
+@ensembles.route("/<path:ensemble_id>/execute", methods=["GET"])
+def execute(ensemble_id):
+    auth = g.get('auth_token')
+    if auth is None:
+        return Response('[]', mimetype = 'application/json')
+    with Session(engine) as session:
+        stmt = select(Ensemble).where(Ensemble.user_id == auth['sub']).where(Ensemble.id == ensemble_id)
+        ensembles = session.scalars(stmt).fetchall()
+        if len(ensembles) == 0:
+            return Response('No such scenario', status = 400)
+        return create_jobs(ensembles[0], auth)
+
+def create_jobs_for_config(config):
```

Comment:
At this point it would make sense to configure various ways of sampling the scenarios (maybe I'll add another issue for that?). As in, users can select if they
- want to create a full factorial sample (sample across the full range of specified parameters)
- go for latin hypercube sampling (even distribition across the parameter space) with a predefined number of scenarios
- go for monte carlo sampling 
etc.

From a logic point of view a quite large number of sampling techniques is already in the exploratory modelling and analysis (EMA) workbench python package (https://emaworkbench.readthedocs.io/en/latest/ema_documentation/em_framework/samplers.html#ema_workbench.em_framework.samplers.LHSSampler) and I think strategically I'd make sense to implement the EMA workbench as part of the UMP for the creation of scenarios

Reply:
That would be possible of course, the question is what the UI for this should look like, perhaps that needs to be specified first?

Reply:
Please also compare: https://github.com/citysciencelab/scenario-explorer-addon/pull/29